### ISSUE TITLE
iOS: AudioCache doesn't play

### ISSUE THREAD

**Issue by haarts** [luanpotter/audioplayer#32]
I'm using `AudioCache` to play a small sound effect when a particular widget loads. This works on Android but no sound is produced on iOS.

The logs (with debugging on) actually show that a sound is being played (or at least I think that is what the logs say):

```
iOS => call play, playerId c0c5f653-24e0-410b-9b7d-819b9e93f824
play!
isLocal: 1 1
volume: 1.000000 1
setUrl /var/mobile/Containers/Data/Application/398C4AF5-447B-4BA5-8652-4EF82010AE54/Library/Caches/bell.mp3
ios -> onTimeInterval...
asdff c0c5f653-24e0-410b-9b7d-819b9e93f824 - 0
asdff end
<snip>
```

The relevant code:

```dart
class Sheep extends StatelessWidget {
  static AudioCache player = AudioCache(prefix: 'sounds/');
  
  @override
  Widget build(BuildContext context) {
    player.play('bell.mp3');

    // some other widget building stuff eventually returning a widget
  }
}
```

Log file: https://pastebin.com/FTUCMd8W

audioplayer version 0.7.5
iOS version 11.4.1
Flutter 0.5.1 • channel beta • https://github.com/flutter/flutter.git
Framework • revision c7ea3ca377 (3 months ago) • 2018-05-29 21:07:33 +0200
Engine • revision 1ed25ca7b7
Tools • Dart 2.0.0-dev.58.0.flutter-f981f09760

**Comment by luanpotter** [comment]
Hi, @haarts,

I'm not sure why this is happening, but I have a few comments that might help:

* Did you try to play using the regular AudioPlayer directly? Is it a problem with the way audio_cache does things or with audioplayers in general?

* I normally add some basic stuff that you probably know but that are easy to miss: did you add the file to your pubspec and run a full 'pub get'? Did you bump the volume (there are several volumes, make sure to change the one that is actually being used)? Is your problem the same as [this one](https://github.com/luanpotter/audioplayer/issues/16)? It's a very cryptic error we could not reproduce where the audio is only played through headphones.

* I believe you shouldn't call play on build. While it might not pose a problem in your case, because it's a stateless component, I believe build might be called other times for any reason Flutter might see fit; you could use methods like the constructor or initState and dispose to start and stop playing audio when the component appears on screen. That's most likely not the cause of your problem, just a discussion about the Flutter lifecycle (and it's just my view about it, it might be wrong!).

Apart form that, I believe @feroult can try to help out, as he has more knowledge about iOS stuff :))

**Comment by mjordan56** [comment]
I believe this could be the same problem reported by timgrubb. See my comment in issue #33

**Comment by feroult** [comment]
Hi guys, thanks for contributing...

Do you have this problem when running on iOS simulator or only on a real device?

Any specific device model?

I'll check it out soon but this info can help a lot :)

**Comment by mjordan56** [comment]
Hi feroult,

Please read my comment in issue #33, I give a lot more information about this bug and the testing I performed.

To answer your question, this bug appears to only happen on a physical device. When I run on the Simulator using an iPhone X the audio sound plays as expected. When I run on a physical iPhone X device the audio plays through the headphones. I believe timgrubb has identified the solution to the problem in his reference on Stackoverflow https://stackoverflow.com/questions/50616323/flutter-apps-audio-not-working-through-speaker.

Michael

**Comment by lclc** [comment]
Thanks a lot @mjordan56.

@feroult: Our test device is also a iPhone X. We don't have another physical device to test available.

**Comment by feroult** [comment]
@mjordan56 I was aware of your info, its pretty good. Also the SO reference.

@lclc are you also getting this error? 

Well, we have an iPhone X in the office, I can check it out there.

Thanks again for the input.

Fer

**Comment by lclc** [comment]
@feroult Yes, I work with Harm and I'm the guy who tests and builds for iOS.

**Comment by luanpotter** [comment]
Fixed by #41. Thanks @tsun424!
